File: dbuf-gen/tests/kotlin/canon/nat_vec.kt
sealed class Nat {
    private constructor() {
        // constructor asserts
    }
    class Suc: Nat {
        val pred: Nat;
        constructor(pred: Nat): super() {
            // inner class asserts
            this.pred = pred;
        }
    }
    class Zero: Nat {
        constructor(): super() {
            // inner class asserts
        }
    }
}
sealed class Vec {
    val n: Nat;
    private constructor(n: Nat) {
        // constructor asserts
        this.n = n;
    }
    class Cons: Vec {
        val value: Nat;
        val tail: Vec;
        constructor(p: Nat, value: Nat, tail: Vec): super(Nat.Suc(p)) {
            // inner class asserts
            this.value = value;
            this.tail = tail;
        }
    }
    class Nil: Vec {
        constructor(): super(Nat.Zero()) {
            // inner class asserts
        }
    }
}

File: dbuf-gen/tests/swift/canon/nat_vec.swift
import Foundation

public enum nat {
    public enum deps {}

    public indirect enum Body: Codable {
        case suc(pred: Nat)
        case zero
    }

    public struct Dependencies: Codable {
    }

    public struct Nat: Codable {
        public var body: Body
        public var dependencies: Dependencies

        public static func suc(pred: Nat) -> Nat {
            let body = Body.suc(pred: pred)
            let dependencies = Dependencies()
            return Nat(body: body, dependencies: dependencies)
        }

        public static func zero() -> Nat {
            let body = Body.zero
            let dependencies = Dependencies()
            return Nat(body: body, dependencies: dependencies)
        }

        public func serialize() -> Data {
            return try! JSONEncoder().encode(self)
        }

        public static func deserialize(_ data: Data) throws -> Nat {
            return try JSONDecoder().decode(Self.self, from: data)
        }
    }
}

public typealias Nat = nat.Nat

public enum vec {
    public enum deps {}

    public indirect enum Body: Codable {
        case cons(value: Nat, tail: Vec)
        case nil
    }

    public struct Dependencies: Codable {
        public var n: Nat
    }

    public struct Vec: Codable {
        public var body: Body
        public var dependencies: Dependencies

        public static func cons(p: Nat, value: Nat, tail: Vec) -> Vec {
            let body = Body.cons(value: value, tail: tail)
            let dependencies = Dependencies(n: Nat.suc(pred: p))
            return Vec(body: body, dependencies: dependencies)
        }

        public static func nil() -> Vec {
            let body = Body.nil
            let dependencies = Dependencies(n: Nat.zero())
            return Vec(body: body, dependencies: dependencies)
        }

        public func serialize() -> Data {
            return try! JSONEncoder().encode(self)
        }

        public static func deserialize(_ data: Data) throws -> Vec {
            return try JSONDecoder().decode(Self.self, from: data)
        }
    }
}

public typealias Vec = vec.Vec

